gnowledge/gstudio
# Newer file storage Implementation on filesystem with Filehive module

**Need of change of file storage Implementation:**
> Previously, we were using mongodb's gridfs to store files. But because of following issues in gridfs, we decided to move to traditional approach of saving files on file system.
 - Indexing error. 
 - Taking dump of chunks.
 - Restoring dump of chunks, mostly gives and error.
 - Serving video from specific position problem.

File saving has been implemented with  **File Hive** module.

--

**Filehive module**
Using pip level dependency of [hashFS](https://pypi.python.org/pypi/hashfs/) for managing files which is content-addressable file management system.
 > HashFS supports basic file storage, retrieval, and removal as well as some more advanced features like file repair.
 
- For the same, defined **`MEDIA_ROOT`** in `settings.py`.
- Instantiated hashFS with `MEDIA_ROOT`.
- All the files will be stored inside this directory under folder `media`.
- We can have `media` folder anywhere in the system (specify `MEDIA_ROOT` accordingly).
- HashFS manages a directory - `media` where files are saved based on the file’s hash.
- Defined new class/collection in `models.py` named: `Filehive`, to store uploaded file's information.
- After putting file inside directory `media`, new `Filehive` instance will be created (for non-duplicate file). Which is holding `md5`, relative path and nine other fields.
- Most probably, Filehive objects will not be updated.

--

**Updated `GSystem` class:**
- Currently, except `File` instances all leaf node instances are of class `GSystem`.
- That's why with newer/modified class structure, incorporated file instances required field.
- Following is added field:
  <pre>
'if_file': {
			'mime_type': basestring,
			'original' : {'id': ObjectId, 'relurl': basestring},
			'mid'      : {'id': ObjectId, 'relurl': basestring},
			'thumbnail': {'id': ObjectId, 'relurl': basestring}
		},
  </pre>
- Along with *original*, considered case for *mid* and *thumbnail*'s too. So that all the necessary info will remain with object itself.
- After file gets saved/put under folder `media`, hashFS returns relative url which will be stored in field `relurl` of respective key (`original`/`mid`/`thumbnail`).

--

**Additional model level changes:**
- `Class Node`:
  - New method: `fill_node_values()`
    - This method can be called with Node instance (object inheriting Node class).
    - We can update Node values either by passing `kwargs` or by passing `request`. And this method will update the current object accordingly.
  - Added code to clean class structure by removing invalid fields (sometimes mongokit produces invalid class structure).
- `class GSystem:`
  - Added new class level static variables:
    1. `image_size_name`
    2. `image_sizes`
    3. `sys_gen_image_prefix`
  - As stated above, added new field: `if_file`.
  - New method: `def fill_gstystem_values()`
    - This method needs to be called with GSystem object.
    - Method will call node's `fill_node_values()` and returns values updated GSystem object.
    - This method takes care of uploading file.
      - If file is image file, it will call methods written in `Filehive` class to convert image to mid and thumbnail size.
      - Also updates related `if_file` field with respective values of those keys (`original`, `mid`, thumbnail).
  - New method: `get_gsystem_mime_type()`
    - Returns `mime_type` of file.
    - Also take care of existing File object structure.
  - New method: `get_file()`
    - Returns file from `md5` or `relative url` argument.
  - Made `GSTUDIO_DEFAULT_LICENSE` as default value for GSystem instances.
- `class Filehive()`:
  - New methods
    - `get_file_mimetype()`
    - `get_file_name()`
    - `get_file_extension()`
    - `convert_image_to_size()`
- Added new Filehive collection variable: `filehive_collection`
- Class related, frequently required methods has been written in respective class to accomplish following:
  - Avoid redundancy by rewriting same code in various views and methods.
  - This will act as class helper function(s).
  - We just need to implement business logic in view's by using these methods.
  - Single point of correction and update. 

--

**Other changes:**
- Added `media/` in `.gitignore` file.
- Added `hashfs` dependency in `requirement.txt` file.
- Updated `sync_existing_documents` to resemble all GSystem instances with current structural changes.
- New templates (may be test template/s):
   - `filehive.html` (copied from uploadDoc.html) to test and upload files in filehive.
   - `filehive_listing.html` to list all the filehive objects.
- Made changes in `ggalleypage.html`, `simple_card.html`, `gcourse.py` and `group.py` to upload files in filehive from raw material and gallery of course.
- Added new urls and view for filehive.
- Modified `settings.py` and `__init__.py` to enable `media` folder accessible.
- Added default license in settings.py: `GSTUDIO_DEFAULT_LICENSE = 'CC-BY-SA 4.0 unported'`
- Modified `readDoc()` in `file.html` to read both filehive and gridfs files.
- Made tags to get saved even after putting tags text (with or without comma separated) in tags input field (node_edit_base.html).
- New `ndf_tag` to get download filename: `get_download_filename`.

--

**TO DO:**
- Run requirements.txt to get `hasfs` (cd till `requirements.txt` file).
  - `pip install -r requirements.txt`
- Making all GSystem instances consistent:
  - `python manage.py sync_existing_documents`

--

**TEST:**
- Upload different types of files in gallery/raw material of course.
- Check uploaded file/s in filehive with following url:
  - `http://<IP>:<port>/home/filehive/list`
- Also you can test upload by following url:
  - `http://<IP>:<port>/home/filehive/upload_form/`
- By default, files will get stored at: `gnowsys-ndf/gnowsys_ndf/ndf/static/media`
  - Try to change media folder's position and change/update it's path accordingly in `MEDIA_ROOT` of settings.py file.
  - Now upload or try to access files.
- Perform create, read and update operation on filehive file instances.
- Add comma separated tags text and click on save button. Check if all the tags got added or not.
